parser-typechecker/src/Unison/CommandLine/InputPatterns.hs

```diff
@@ -1212,8 +1212,8 @@ exactDefinitionOrPathArg =
 exactDefinitionQueryArg :: ArgumentType
 exactDefinitionQueryArg =
   ArgumentType "definition query" $
-    bothCompletors (termCompletor exactComplete)
-                   (typeCompletor exactComplete)
+    bothCompletors (termCompletor fuzzyComplete)
+                   (typeCompletor fuzzyComplete)
```

Comment:
Changing `exactDefinitionQueryArg` from `exact` to `fuzzy` seems suspect :)

maybe a `suffixComplete` and a `exactSuffixQueryArg`?

Reply:
For context: we already have a `fuzzyDefinitionQueryArg`, which is distinguished from `exactDefinitionQueryArg` by being `fuzzyComplete` instead of `exactComplete`; it had some [forgotten?] undesirable qualities though, which is why we added `exact`.  Something that dealt in suffixes would still be good though!

Not sure if existing suffixify code would work here?

Reply:
I’m thinking just rename this to ‘definitionQueryArg’. Whether this uses fuzzy or exact or whatever completions more of implementation detail IMO.

FYI fuzzy matching is just substring matching (might be case insensitive). It used to be some bonkers edit distance thing that gave terrible results but we axed that. 😀 

No need for suffix completion imo since a suffix match is also a substring match, and I haven’t yet felt like the substring match is returning too many results. One interesting example - List.map finds base.List.map.doc which seems nice to me. 😀

Reply:
Suspect the separation was because the fuzzy matching was overly fuzzy at one time, so we didn’t want to use it everywhere. 😀

It’s safe to have the completion provider do something sloppy as long as the thing it completes to is a subset of what the command can accept. It’s more that you don’t want to be shown a gajillion completions that aren’t very good. I think the substring matching is pretty good balance.

Reply:
Agree we don't want completions that aren't very good; we do want auto-complete to let the user make progress to their desired identifier as quickly as possible.  I'm thinking it's going to do the former though.

Could you do an audit of what was using fuzzy match (if anything) and what was using exact match, and make an inventory (here I guess) of what's going to be affected by the change?  Or make a smaller change and hold off on the comprehensive analysis for now.

Reply:
But eventually (or now) we should make an taxonomy of the types and purposes of arguments, and have a consistent way of decoding and auto-completing them.  Within the implementation for each identified argument type, we can have changeable implementation details! /cc @runarorama

Reply:
Okie dokie, did that renaming to `definitionQueryArg` and you can peep [this diff](https://github.com/unisonweb/unison/pull/1371/commits/cda53566b51366a931203e983d5401e51f84e18d) to see the commands whose completions will be affected by this PR. I think these are all very reasonable spots to do fuzzier completion.

> It’s safe to have the completion provider do something sloppy as long as the thing it completes to is a subset of what the command can accept.

This is important to understand, can you ack? Basically, "be liberal in what you autocomplete, but strict in what you accept" Example is if `names List.map<tab>` provides completions of `base.List.map` and `base.List.map.doc` - the user still ultimately has to pick which one they want, they are just shown more options than the exact (really prefix) matching.

I basically don't think prefix-only completion is super useful, for this reason (exception is maybe `ls` and `cd` commands). Something we can consider for later is doing a sort / cutoff if there's too many results, and then we can develop some interesting criteria for doing the sorting. Like if we really think prefix matches should come before others, we could do that.

Reply:
> It’s safe to have the completion provider do something sloppy as long as the thing it completes to is a subset of what the command can accept.

and
 
> Basically, "be liberal in what you autocomplete, but strict in what you accept"

I suppose there's a spectrum for what this could mean, but at face value, I disagree — too many sloppy completions can make an auto-complete feature worse than no auto-complete feature. e.g. You can lose time looking through completions that aren't what you want, before giving up; all cost, no benefit. So I think it's important to be thoughtful about how to give a good set of completions.

I'm not implying that you haven't already been thoughtful and have reached a great solution (and I myself definitely haven't), just that it's not clear from the PR description and diff alone!  Agree that the completions you suggested in the `List.map` example sound good.

Reply:
Yeah I think we agree that if autocomplete shows mostly nonsense that’s worse. 😀 Basically I think we are on the same page. 

Autocomplete provider is easy to futz with later, too so we can easily revisit the matching later once we have more experience. IMO it’s a big improvement.

Are you ok with closing this out for now?

Reply:
Do you have a sense of whether this will create a feeling of inconsistency relative to whatever arg. types are still using `exactComplete`?  Or are those ones that should stay as-is?

Anyway, we can merge this and see what happens.

Reply:
Created #1377 to track taking a look at other command args.